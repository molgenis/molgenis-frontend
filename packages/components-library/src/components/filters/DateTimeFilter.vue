
<template>
  <b-input-group class="flex-nowrap">
    <date-range-picker
      ref="picker"
      v-model="dateRange"
      class="flex-grow-1"
      :opens="opens"
      :locale-data="{ firstDay: 1, format: pickerFormat }"
      :single-date-picker="!range"
      :time-picker="time"
      :time-picker24hour="true"
      :show-week-numbers="false"
      :show-dropdowns="true"
      :auto-apply="false"
      :ranges="false"
      :linked-calendars="false"
      :always-show-calendars="true"
      :min-date="min"
      :max-date="max"
      :append-to-body="appendToBody"
      @update="updateValues"
    >
      <template slot="input">
        {{ formattedDate }}
      </template>
    </date-range-picker>
    <b-input-group-append>
      <b-button
        variant="outline-secondary"
        class="t-btn-clear"
        @click="clearValue"
      >
        <font-awesome-icon icon="times" />
      </b-button>
    </b-input-group-append>
  </b-input-group>
</template>

<script>
export default {
  name: 'DateTimeFilter',
  props: {
    /**
     * Define the start of your calendar range.
     * @values ISOString, null
     */
    min: {
      default: () => null,
      type: [String, null]
    },
    /**
     * Define the end of your calendar range.
     * @values ISOString, null
     */
    max: {
      default: () => null,
      type: [String, null]
    },
    /**
     * Defines a maximal Date/Datetime thats selectable in the calendar widget.
     * @values right, left, top, bottom
     */
    opens: {
      default: () => 'right',
      type: String
    },
    /**
     * Toggles time selection in the calendar widget.
     * @values true, false
     */
    time: {
      type: Boolean,
      default: () => true
    },
    /**
     * Toggles single date selection; uses only min property in that case.
     * @values true, false
     */
    range: {
      type: Boolean,
      default: () => true
    },
    /**
     * The default selected date.
     * @model
     */
    value: {
      type: Array,
      default: () => [null, null]
    },
    /**
     * Append the dialog to the body tag, useful if filter is used within a container that uses overflow:scroll
     * @model
     */
    appendToBody: {
      type: Boolean,
      default: () => false
    }
  },
  data: function () {
    return {
      dateRange: {
        startDate: null,
        endDate: null
      }
    }
  },
  computed: {
    formattedDate: function () {
      const date = this.dateRange
      if (!date.startDate || !date.endDate) {
        return 'Select...'
      }
      if (date.startDate.toISOString() === date.endDate.toISOString()) {
        return this.formatDate(date.startDate)
      } else {
        return `${this.formatDate(date.startDate)} - ${this.formatDate(date.endDate)}`
      }
    },
    pickerFormat () {
      return this.time ? 'yyyy-mm-dd HH:MM:ss' : 'yyyy-mm-dd'
    }
  },
  watch: {
    value (newValue) {
      this.setDateRange(newValue)
    }
  },
  beforeMount () {
    this.setDateRange(this.value)
  },
  methods: {
    clearValue: function () {
      this.dateRange = {
        startDate: null,
        endDate: null
      }

      this.$emit('input', undefined)
    },
    setDateRange (value) {
      if (value) {
        this.dateRange.startDate = this.createDateFromValue(value[0])
        this.dateRange.endDate = this.createDateFromValue(value[1])
      }
    },
    updateValues: function () {
      this.$emit('input', [this.dateRange.startDate, this.dateRange.endDate])
    },
    createDateFromValue (value) {
      if (value) {
        return isNaN(value) ? new Date(Date.parse(value)) : new Date(value)
      }
    },
    formatDate (dateTime) {
      return this.time ? dateTime.toLocaleString() : dateTime.toLocaleDateString()
    }
  }
}
</script>

<style lang="css">
.form-control.reportrange-text {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}
</style>

<docs>
Filter between a begin Date(time) and optionally an end Date(time)

### Usage
```jsx
const defaultMin = new Date()
const defaultMax = new Date()
defaultMax.setDate(defaultMax.getDate() + 7)

const selectMax = new Date()
const selectMin = new Date()
selectMax.setDate(selectMax.getDate() + 30)
selectMin.setDate(selectMin.getDate() - 7)

const model = [defaultMin.toISOString(), defaultMax.toISOString()]
<DateTimeFilter
    opens='right'
    v-bind:max='selectMax.toISOString()'
    v-bind:min='selectMin.toISOString()'
    v-bind:range='true'
    v-bind:time='true'
    v-model='model'>
</DateTimeFilter>
<div>{{model}}</div>
```
</docs>
