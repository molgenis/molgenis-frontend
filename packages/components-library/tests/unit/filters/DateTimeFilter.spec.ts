// @ts-ignore
import { createLocalVue, mount } from '@vue/test-utils'
import DateTimeFilter from '@/components/filters/DateTimeFilter.vue'
import { BootstrapVue } from 'bootstrap-vue'

const localVue = createLocalVue()
localVue.use(BootstrapVue)

const buildWrapper = (propsData: any) => {
  const wrapper = mount(DateTimeFilter, {
    listeners: {
      input: (newVal: any) => {
        wrapper.setProps({ value: newVal })
      }
    },
    localVue,
    propsData
  })

  return wrapper
}

describe('Datetimefilter.vue', () => {
  describe('in date time mode', () => {
    const wrapper = buildWrapper({
      name: 'datetime',
      label: 'Datetime',
      collapsed: false,
      max: null,
      min: null,
      opens: 'right',
      range: true,
      time: true
    })
    const vm: any = wrapper.vm

    it('sets a date range', async () => {
      await wrapper.find('.vue-daterange-picker div').trigger('click')
      await wrapper.findAll('.yearselect').setValue('2020')

      await wrapper.find('.drp-calendar.left .monthselect').setValue('3')
      await wrapper.find('.drp-calendar.left .monthselect').setValue('3')

      await wrapper.find('.vue-daterange-picker div table tr:nth-child(2) td:nth-child(7)').trigger('click')
      await wrapper.find('.vue-daterange-picker div table tr:nth-child(4) td:nth-child(1)').trigger('click')
      await wrapper.find('.applyBtn').trigger('click')

      expect(vm.dateRange.startDate.toISOString()).toBe('2020-03-01T12:00:00.000Z')
      expect(vm.dateRange.endDate.toISOString()).toBe('2020-03-09T12:00:00.000Z')
      expect(wrapper.emitted().input).toBeTruthy()
      await wrapper.find('.t-btn-clear').trigger('click')
      expect(vm.dateRange.startDate).toBe(null)
      expect(vm.dateRange.endDate).toBe(null)
    })

    describe('when createDateFromValue is passed a string value', () => {
      const parsed = vm.createDateFromValue('04 Dec 1995 00:12:00 GMT')
      it('should parse the string to date', () => {
        expect(parsed).toEqual(new Date(Date.parse('04 Dec 1995 00:12:00 GMT')))
      })
    })
  })

  describe('in date only mode', () => {
    const wrapper = buildWrapper({
      name: 'datetime',
      label: 'Datetime',
      max: null,
      min: null,
      range: true,
      time: false
    })
    const vm: any = wrapper.vm

    it('formatDate should leave of the time part', async () => {
      const date = new Date(Date.parse('04 Dec 1995 00:12:00 GMT'))
      expect(vm.formatDate(date)).toEqual(date.toLocaleDateString())
    })
  })
})
