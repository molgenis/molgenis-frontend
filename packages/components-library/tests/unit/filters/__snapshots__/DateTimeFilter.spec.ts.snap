// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datetimefilter.vue matches the snapshot 1`] = `
<div
  class="input-group flex-nowrap"
  label="Datetime"
  name="datetime"
  role="group"
>
  <!---->
  <div
    class="vue-daterange-picker flex-grow-1"
    data-v-267f4ee2=""
  >
    <div
      class="form-control reportrange-text"
      data-v-267f4ee2=""
    >
      
      Select...
    
    </div>
    <transition-stub
      data-v-267f4ee2=""
      mode="out-in"
      name="slide-fade"
    >
      <!---->
    </transition-stub>
  </div>
   
  <div
    class="input-group-append"
  >
    <button
      class="btn t-btn-clear btn-outline-secondary"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-times fa-w-11"
        data-icon="times"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 352 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
  <!---->
</div>
`;
